// skedyul/dist/core/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.communicationChannel = exports.workplace = void 0;
const CORE_BASE = process.env.SKEDYUL_NODE_URL ?? '';
async function callCore(method, params) {
    const response = await fetch(`${CORE_BASE}/core`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ method, params }),
    });
    const payload = await response.json();
    if (!response.ok) {
        throw new Error(payload?.error?.message ?? `Core API error (${response.status})`);
    }
    return payload;
}
exports.workplace = {
    async list(args) {
        const payload = await callCore('workplace.list', args?.filter ? { filter: args.filter } : undefined);
        return payload.workplaces;
    },
    async get(id) {
        const payload = await callCore('workplace.get', { id });
        return payload.workplace;
    },
};
exports.communicationChannel = {
    async list(filter) {
        const payload = await callCore('communicationChannel.list', filter ? { filter } : undefined);
        return payload.channels;
    },
    async get(id) {
        const payload = await callCore('communicationChannel.get', { id });
        return payload.channel;
    },
};
