// skedyul@0.1.2/dist/core/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.coreApiService = void 0;
class CoreApiService {
    register(service) {
        this.service = service;
    }
    getService() {
        return this.service;
    }
    setWebhookHandler(handler) {
        this.webhookHandler = handler;
    }
    async dispatchWebhook(request) {
        if (!this.webhookHandler) {
            return { status: 404 };
        }
        return this.webhookHandler(request);
    }
    async callCreateChannel(channel) {
        return this.service?.createCommunicationChannel(channel);
    }
    async callUpdateChannel(channel) {
        return this.service?.updateCommunicationChannel(channel);
    }
    async callDeleteChannel(id) {
        return this.service?.deleteCommunicationChannel(id);
    }
    async callGetChannel(id) {
        return this.service?.getCommunicationChannel(id);
    }
    async callListChannels() {
        return this.service?.getCommunicationChannels();
    }
    async callGetWorkplace(id) {
        return this.service?.getWorkplace(id);
    }
    async callListWorkplaces() {
        return this.service?.listWorkplaces();
    }
    async callSendMessage(args) {
        return this.service?.sendMessage(args);
    }
}
exports.coreApiService = new CoreApiService();
